docs: restructure concise evaluating guide,  add terminology list

janderssonse opened the pull request:
This restructure suggestion PR does the following:

The most important:

* Restructures the information (and does not remove any information from the original document) into simple tables with a theme header, and rules-description format. Aims to make the structure for the user clearer, and hit a add sweet spot between between, let's say a formal ISO 270x doc and something aimed at a developer/tech doing assessment.
* In the same mood, tightens up the language making it a bit more imperative, Verify that, Do that, ex.!

Other:
* Adds basic terminology list
* Adds a version table
* Adds a checkbox to that tables, (This is just an idea, and I don't know if I even like it myself, maybe not... let me know). Intended to use as a check-list, for the end user. Stay in, remove it?

This suggestion came from another PR where I asked about translations, and, while translating, I thought I would like to make the structure a bit more clearer in the original document.

david-a-wheeler commented:
Definitely prettier! It's also _much_ longer (e.g., when printed). I don't know which is better for people. Does anyone else have any thoughts?

janderssonse commented:
Thanks. The terminology list could be removed as it takes up a lot of the new document. Maybe a terminology list could be an idea for a separate md-doc (in the future). But as for now, I could just remove it from the suggestion above.

gkunz commented on docs/Concise-Guide-for-Evaluating-Open-Source-Software.md:
```diff
@@ -1,57 +1,117 @@
+| **Dependency** | External code packages or libraries that a software project relies on to function properly. |
+| **LTS (Long Term Support)** | A version of software that will receive maintenance and security updates for an extended period, often used for stability in production environments. |
+| **OpenChain Security Assurance Specification** | A specification from the [OpenChain Project](https://www.openchainproject.org/) (part of the Linux Foundation) that provides guidelines for managing security assurance in the open source supply chain. It helps organizations systematically identify and remediate known vulnerabilities in open source components. While OpenChain's license compliance  ...
+| **OpenSSF Badge** | A certification from the [Open Source Security Foundation](https://openssf.org/) that indicates a project follows security best practices. Projects earn badges by meeting specific criteria for security, quality, and community engagement through the [Best Practices Badge Program](https://www.bestpractices.dev/). |
```
To complement my suggestion above to include the new Security Baseline:

```suggestion
| **OpenSSF Badge** | A certification from the [Open Source Security Foundation](https://openssf.org/) that indicates a project follows security best practices. Projects earn badges by meeting specific criteria for security, quality, and community engagement through the [Best Practices Badge Program](https://www.bestpractices.dev/). |
| **OpenSSF Open Source Security Baseline** | A [collection of secure development requirements](https://baseline.openssf.org/) for open source projects relative to their maturity level. . |
```

janderssonse replied:
Added as suggested.

gkunz reviewed (commented):
I in fact like the new layout, the check boxes and the languages changes. I think that the increase in length is outweighed by the more prominent structure. However, at the end of the day, it is a matter of taste.

Independent of the structure, I made two proposed additions to the content: the Security Baseline and API Stability as a new criterion. Probably we shouldn't mix structural changes and content changes, so I am happy to add them independently to the guide. Still, curious about feedback.

janderssonse commented:
Note: If you decide you want the checkbox to stay in, I will change it to '[ ]' instead of the current Unicode character □. The '[ ]' format is commonly used in markdown and makes it easy to tick off items, ie not just for printout. I was in a bit of a hurry when I added the Unicode character column - I should have used '[ ]' from the start.

agreedSkiing commented:
Colleague of @janderssonse chiming in. I would recommend to use "real" checkboxes `| <ul><li> [ ] </li></ul>  |` instead of `□` if they work in githubs markdown generator. Credit for this goes to https://stackoverflow.com/a/47344640

gkunz commented:
Thanks! As the guide is rendered to https://best.openssf.org/Concise-Guide-for-Evaluating-Open-Source-Software, do we know what the output of `| <ul><li> [ ] </li></ul>  |` looks like there?

janderssonse commented:
Regarding Checkbox rendering:

I did some further experimentation and investigated both GitHub flavored markdown and the kramdown conversions (the rendering of the markdown in the link above is done with Jekyll, which uses kramdown GFM). I came to the conclusion that a Unicode character is the only solution that works for both cases I.E- rendering with GitHub's native renderer directly, and through Jekyll/Kramdown.

The official GitHub checkbox syntax - [ ] and above suggestion will render a checkbox in GitHub, but not in Jekyll/Kramdown.
Similarly, using an HTML input element would work for Jekyll/Kramdown, but gets removed from the GitHub native rendering as it sanitizes HTML.

**So, to summarize,  sadly, the best approach to support both cases (GitHub native rendering and Jekyll-kramdown-GFM rendering) is to use a Unicode character checkbox.**

Note:
I confirmed this through local conversion testing with: kramdown --input GFM withcheckboxvariants.md > a.html
(and AI would reach the same conclusion when provided with sufficient context).

gkunz commented:
Thanks @janderssonse for looking into this. I also confirmed the same result in the meantime.

maennchen commented on docs/Concise-Guide-for-Evaluating-Open-Source-Software.md:
```suggestion

**Document Version History**
```

Heading Is rendered as a table row:

![image](https://github.com/user-attachments/assets/76b1db86-f4b7-4268-bcc3-bf91c8930193)

janderssonse replied:
Fixed the header as noted

janderssonse commented:
> We don't need the version number, the date is sufficient. We normally only use dates.

Fixed, versions removed, only using date

janderssonse commented on docs/Concise-Guide-for-Evaluating-Open-Source-Software.md:
```diff
@@ -1,57 +1,117 @@
+| **Dependency Management** | Verify that the package dependencies are (relatively) up to date. | □ |
+| **Repository Security** | Confirm that developers use code hosting security features where applicable (e.g., if they're on GitHub or GitLab, do they use branch protection). | □ |
+| **Security Audits** | Identify existing security audits and verify that identified vulnerabilities have been fixed. Security audits are relatively uncommon, but see OpenSSF's "[Security Reviews](https://github.com/ossf/security-reviews)". | □ |
+| **Security Development** | Confirm they apply many practices in the [Concise Guide for Developing More Secure Software](https://best.openssf.org/Concise-Guide-for-Developing-More-Secure-Software). | □ |
```
Added as suggested.

janderssonse commented on docs/Concise-Guide-for-Evaluating-Open-Source-Software.md:
```diff
@@ -1,57 +1,117 @@
+
+| Rule | Description | Checkbox |
+|------|-------------|:--------:|
+| **Interface Design** | Verify that the interface/API is designed to be easy to use securely (e.g., if the interface implements a language, does it support parameterized queries). | □ |
```
Added as suggested.

janderssonse commented on docs/Concise-Guide-for-Evaluating-Open-Source-Software.md:
```diff
@@ -1,57 +1,117 @@
+|------|-------------|:--------:|
+| **License Clarity** | Verify that every component has a license, that it's a widely-used [OSI license](https://opensource.org/licenses) if it's OSS, and that it's consistent with your intended use. Projects that won't provide clear license information are less likely to follow other good practices that lead to secure software. | □ |
+| **Name Verification** | Check if a similar name is more popular - that could indicate a typosquatting attack. | □ |
+| **Usage Metrics** | Assess if it has significant use. Software with many users or large users may be inappropriate for your use. However, widely-used software is more likely to offer useful information on how to use it securely, and more people will care about its security. | □ |
```
Yes, I just used the original. I also think the sentence is not clear, and even that it should be clearer without that sentence. '...use. Widely-used....'

david-a-wheeler replied:
I was thinking about the point of [You are not Google](https://blog.bradfieldcs.com/you-are-not-google-84912cf44afb): "if you’re using a technology that originated at a large company, but your use case is very different, it’s unlikely that you arrived there deliberately; no, it’s more likely you got there through a ritualistic belief that imitating the giants would bring the same riches."

Basically, while software with many users is often capable, it *might* not be a good match for your use case. You need to choose software relevant to your problem, not just choosing some software because it's the latest fad. The text is an attempt to counter fad-driven development aka [Hype Driven Development](https://blog.daftcode.pl/hype-driven-development-3469fc2e9b22).

I'm quite open to finding another way to say it.

david-a-wheeler replied:
I think this should be a separate issue/PR. Let's not hold up the restructure for rewording this one point. This is the same wording as earlier.

janderssonse replied:
Thanks for clarifying, I understand better now. While I understand and agree to the philosophical reasoning -  , I wonder if a reader would catch that, without a longer explanation. An idea would be to put the sentence under the  "Consider Necessity" question, and in a more straightforward form?  Something like -> Does your use case really motivate the addition, or are there simpler solutions? 

But, maybe the whole discussion as you suggest, should be moved to another issue.

janderssonse commented:
I have rebased, and fixed most of the suggestions. Questions still:
- should checkbox stay or go?
- should terminology list stay or go?
- should the unclear sentence @gkunz mentions further up under "Usage Metrics" stay or be trimmed?

david-a-wheeler commented:
My thoughts:

> should checkbox stay or go?

If we can get the checkboxes to "work" online they have some value. We use [Kramdown with Jekyll](https://jekyllrb.com/docs/configuration/markdown/) (the default for GitHub Pages). We can modify its configuration, but is there a configuration that makes them "work"? I'm not sure the Unicode checkboxes add much value, but it doesn't matter much to me either way.

> should terminology list stay or go?

I think it should stay. Having clear terms' definitions seems useful. Having them at the *end* means people can see the meat first.

> should the unclear sentence @gkunz mentions further up under "Usage Metrics" stay or be trimmed?

You mean "Software with many users or large users may be inappropriate for your use." yes? Let's make that a separate PR. I don't want to hold up this reorg while also trying to rewrite the text, and trying to do too many things at once can be confusing.

david-a-wheeler commented:
[janderssonse](https://github.com/janderssonse) found that Kramdown can't be configured to make this work. At first I thought we could configure Kramdown to make the checkboxes work ("that can't be right!") but I sure don't find a solution either :-(.

david-a-wheeler commented:
I found this which uses JavaScript to fix up some things. I'm not sure this effort is worth it: https://thatgurjot.com/til/clickable-checkboxes-on-jekyll-hugo/

david-a-wheeler commented:
In the long term we're working on an OpenSSF glossary. Once that has real content we could point off to that for the terminology.

janderssonse commented:
> [janderssonse](https://github.com/janderssonse) found that Kramdown can't be configured to make this work. At first I thought we could configure Kramdown to make the checkboxes work ("that can't be right!") but I sure don't find a solution either :-(.

As the checkbox currently seems to be of limited use, and is  causing problems for different renderings options..
I tried another idea - removing the checkbox character, leaving the table cell, with the header "Check" instead of checkbox. Still gives the impression to the reader that one should work through this guide and fill in the blanks. 

Next option would be that we just remove the Check-cell totally.

janderssonse commented:
Rebased. Any chance of seeing this move forward? :)

What is left is, from what I can
- make a decision to leave an empty cell named check, which avoids any rendering problems for checkbox (done)
- make a decision about the "usage metrics" discussion earlier in the thread.

gkunz commented:
Hi @janderssonse. Sorry for the slow response times (Kubecon and other things happened). I support moving this forward and focusing this PR on restructuring the content. Specifically:

* I don't mind having no functioning checkboxes, but I think that adding a "check" column provides value by explicitly encouraging treating this as a checklist. I personally would have called the header "checked" instead of "check" (mainly because at first reading, I was expecting to find a check in this column) but this is very close to bike-shedding.
* let's move my feedback regarding some of the content to another PR - I'll take the action to propose new text once this PR here lands
* I also agree with keeping the glossary for now and move it later if needed.

gkunz reviewed (approved):
LGTM. I propose to move this forward.

david-a-wheeler commented on docs/Concise-Guide-for-Evaluating-Open-Source-Software.md:
```diff
@@ -1,57 +1,117 @@
+
+- [The Tidelift guide to choosing packages well (February 2021)](https://tidelift.com/subscription/choosing-open-source-packages-well), Tidelift
+- [How to Evaluate Open Source Software / Free Software (OSS/FS) Programs](https://dwheeler.com/oss_fs_eval.html)
+
+
```
```suggestion

```

You need to delete one of the lines to pass markdownlint.

janderssonse replied:
Should be fixed now

david-a-wheeler replied:
Great! Which blank line did you delete :-) :-) ?

janderssonse commented:
>  I notice that the markdowlinter complains so will fix that, and also change to your suggestion "Checked". I'll fix it later.

Both of them should now be fixed

david-a-wheeler commented:
To continue the bikeshedding, maybe `Checked` should say `Done?` - it's shorter and I think that's the intent.

janderssonse commented:
> To continue the bikeshedding, maybe `Checked` should say `Done?` - it's shorter and I think that's the intent.

Done... is Done!:)

gkunz commented:
Looks good to me. If we get another approving review, I'll merge this PR.

janderssonse commented:
> The new alphabetic organization seems to suggest that consulting [SAFECode’s guide _Principles for Software Assurance Assessment_](https://safecode.org/resource-managing-software-security/principles-of-software-assurance-assessment/) (2019) is the first thing to do. Originally this was point 9, as merely something you COULD do. Do we want to emphasize it that strongly?

How about changing "Utilize established... " to "Consider utililzing established ..." which should make it clear that you could, not should?

janderssonse commented:
> That helps, but s/utilizing/using/ ("utilizing" is an unnecessarily fancy word).

Rebased and utilized! Fixed I meant!

gkunz commented:
thanks @janderssonse. I'd like to really get this merged, so I propose doing so. The WG approved the restructure some time ago already.

Now, one last thing: there is a brand bew OpenSSF global glossary now: https://github.com/ossf/glossary

I suggest moving the terms from this doc to that glossary (I am also fine with first merging and then moving things around) @david-a-wheeler ?

janderssonse commented:
> thanks @janderssonse. I'd like to really get this merged, so I propose doing so. The WG approved the restructure some time ago already.
> 
> Now, one last thing: there is a brand bew OpenSSF global glossary now: https://github.com/ossf/glossary
> 
> I suggest moving the terms from this doc to that glossary (I am also fine with first merging and then moving things around) @david-a-wheeler ?

Thanks, I removed the glossary for now and will add to the global glossary in seperete PR, after this one is done.

david-a-wheeler commented:
Ok. I think we need merge, then improve further, or the bikeshedding will continue forever.
